Std/Data/String/Basic.lean

```diff
@@ -1,15 +1,63 @@
 /-
 Copyright (c) 2022 Jannis Limperg. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Jannis Limperg
+Authors: Jannis Limperg, James Gallicchio, F. G. Dorais
 -/
 
 import Std.Data.Char
 import Std.Data.Nat.Lemmas
+import Std.Data.Array.Match
 
-protected theorem String.Pos.ne_zero_of_lt : {a b : Pos} → a < b → b ≠ 0
+namespace String
+
+instance : Coe String Substring := ⟨toSubstring⟩
+
+protected theorem Pos.ne_zero_of_lt : {a b : Pos} → a < b → b ≠ 0
 | _, _, hlt, rfl => Nat.not_lt_zero _ hlt
 
+/-- KMP Matcher -/
```

Comment:
KMP needs to be explained here (it should be okay on functions as long as the type explains it, but generally you shouldn't assume too much context in docstrings)

Reply:
Actually this is probably a good place to put the overall documentation on how to use this type effectively, maybe put a paragraph and a code block showing basic usage.

Reply:
I found an embarrassing bug so I added a basic test suite as well.

Reply:
What do you think about the description? Should I add something about how the basic working of the algorithm? (I don't think it's necessary for users to know how it works but this is a neat one.)

Reply:
The docs here can just reference the algorithm, possibly including a link to e.g. the wikipedia page. For a more detailed description of the algorithm you should use the module doc on `Data.Array.Match`.